### PR TITLE
Make `Stem dynamics` differentiable/vectorized

### LINKED ISSUE #46: [Task]: Make `Stem dynamics` differentiable and efficient
The module [`Stem dynamics`](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/stem_dynamics.py) has one class `WOFOST_Stem_Dynamics` that should be changed. See explanations and requirements in https://github.com/WUR-AI/diffWOFOST/issues/39.

### PR DESCRIPTION
**Pull request by SCiarella**
Closes #46.

---
Notice that specific test data for the `Stem dynamics` module do **not exist** in `wofost72`, so `test_stem_dynamics.py` implemented in this PR mainly checks the consistency of this new implementation (there are only a few `assert_is_close` statements) .
However, the integration test `test_wofost_pp_with_stem_dynamics` confirms that this new module numerically reproduces the correct results when used in combination with all the other wofost72's modules.

### CODE REVIEW

File: src/diffwofost/physical_models/crop/stem_dynamics.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                (internal and external) state variables. See PCSE documentation for
+                details.
+            parvalues (ParameterProvider): A dictionary-like container holding
+                all parameter sets (crop, soil, site) as key/value. The values are
+                arrays or scalars. See PCSE documentation for details.
+            shape (tuple | torch.Size | None): Target shape for the state and rate variables.
+        """
+        self.kiosk = kiosk
+        self.params = self.Parameters(parvalues)
+        self.rates = self.RateVariables(kiosk, publish=["DRST", "GRST"])
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/stem_dynamics.py]:
```suggestion
        self.params = self.Parameters(parvalues, shape=shape)
        self.rates = self.RateVariables(kiosk, publish=["DRST", "GRST"], shape=shape)
```

File: src/diffwofost/physical_models/crop/stem_dynamics.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+        # INITIAL STATES
+        params = self.params
+        shape = params.shape
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/stem_dynamics.py]:
```suggestion
```

File: src/diffwofost/physical_models/crop/stem_dynamics.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+        # Set initial stem biomass
+        TDWI = params.TDWI
+        FS = self.kiosk["FS"]
+        FR = self.kiosk["FR"]
+        WST = (TDWI * (1 - FR)) * FS
+        DWST = torch.zeros(shape, dtype=self.dtype, device=self.device)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/stem_dynamics.py]:
```suggestion
        DWST = torch.zeros(self.params.shape, dtype=self.dtype, device=self.device)
```

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@SCiarella Thanks! 👍 I left some comments/suggestions. Main issues:
- the sonar reliability issue about `shape`, I wonder if my suggestion fixes the issue, let's see 🤔 
- including output variable `WST` in the gradient tests

If something is unclear, let me know.
